algorithms/java/src/org/openda/algorithms/kalmanFilter/SteadyStateFilter.java

```diff
@@ -208,24 +213,86 @@ public void analysis(IStochObserver observations, IVector obsValues, IVector pre
 
 		int m=innovation.getSize();
 		IVector pred_a = this.mainModel.getObservationOperator().getObservedValues(observations.getObservationDescriptions());
-		for(int i=0;i<m;i++){
+		ArrayList<Integer> missingObservationIndices = new ArrayList<>();
+		ArrayList<Integer> availableObservationIndices = new ArrayList<>();
+		for (int i = 0; i < gainStorageObservationIdsArray.length; i++) {
+			int indexOf = Arrays.binarySearch(obsIds, gainStorageObservationIdsArray[i]);
+			if (indexOf >= 0) {
+				availableObservationIndices.add(i);
+				continue;
+			}
+			missingObservationIndices.add(i);
+		}
+		int numberOfMissingObservations = missingObservationIndices.size();
+		int numberOfAvailableObservations = gainStorageObservationIdsArray.length - numberOfMissingObservations;
+		Matrix m1 = new Matrix(numberOfMissingObservations, numberOfAvailableObservations);
+		Matrix m2 = new Matrix(numberOfMissingObservations, numberOfMissingObservations);
+		for (int rowIndex = 0; rowIndex < missingObservationIndices.size(); rowIndex++) {
+			Integer missingObservationRowIndex = missingObservationIndices.get(rowIndex);
+			for (int columnIndex = 0; columnIndex < availableObservationIndices.size(); columnIndex++) {
+				Integer availableObservationColumnIndex = availableObservationIndices.get(columnIndex);
+				m1.setValue(rowIndex, columnIndex, hk[missingObservationRowIndex][availableObservationColumnIndex]);
+			}
+			for (int columnIndex = 0; columnIndex < missingObservationIndices.size(); columnIndex++) {
+				Integer missingObservationColumnIndex = missingObservationIndices.get(columnIndex);
+				m2.setValue(rowIndex, columnIndex, hk[missingObservationRowIndex][missingObservationColumnIndex]);
+			}
+		}
+		Matrix dAvailable = new Matrix(numberOfAvailableObservations, 1);
+		Matrix identityMatrix = new Matrix(numberOfMissingObservations, numberOfAvailableObservations);
+		for (int i = 0; i < numberOfMissingObservations; i++) {
+			identityMatrix.setValue(i, i, 1);
+		}
+
+		for (int i = 0; i < m; i++) {
 			// find matching column in steady-state gain
-			String gainVectorId = obsIds[i]+":"+Math.round(obsTimeOffsets[i]*24.0*3600.0); //conversion to seconds
-			Results.putProgression("processing obs "+gainVectorId+"\n");
+			String gainVectorId = obsIds[i] + ":" + Math.round(obsTimeOffsets[i] * 24.0 * 3600.0); //conversion to seconds
```

Comment:
This was part of the old code as well. But it puzzles me a bit what this stuff with conversion to seconds is here. You could add a line of comment to explain what kind of ID we are going to construct.